### DISCUSSION TITLE
Git command Cheatsheet

### DISCUSSION THREAD

**Discussion by trang-ngn** [AuDojo/AuDojo#9] [Git]
# Git command cheatsheet

## Basic
```
$ git pull | $ git fetch && git merge
$ git add
$ git commit -m "your message" 
$ git push
```

## Good Practice with git commit message
Commit for small changes with clearly messages => better for readers and your future
```
git commit -m "Add margin" // 😕 Not so helpful
git commit -m "Add margin to nav items to prevent them from overlapping the logo" // 🤩 better
```

## Useful commands for working with branch

| Use/Description | Command |
| ----------------------- | ---------------- |
|  Create a new branch | `$ git branch branch_name` |
|  Switch branch | `$ git checkout main` or `$ git checkout branch_name` |
|  Delete branch (If the branch hasn't been merged, local)| `$ git branch -D branch_name ` |
|  Delete branch (remotely) | `$ git push origin --delete branch_name` |
|  Temporarily saves your uncommitted changes so you can work on something else without committing (e.g. switch to another branch but don't want to commit changes) |`$ git stash` |
|  Applies the most recent stashed changes and removes them from the stash list. | `$ git stash pop` |
|  Removes all stashed entries. | `$ git stash clear` |
|  Takes the changes from the specified commit and applies them to the top of current branch. (copy commit_hash from `git log`) | `$ git cherry-pick <commit_hash>` |
|  Shows a detailed list of commits | `$ git log` or `$ git log --oneline` |
|  Updates the commit message of the last commit. (e.g. when you forgot sth like removing whitespaces and it is not necessary to create a new commit) | `$ git commit --amend -m "your new message"` and then `$ git push --force` |
|  Update your feature branch with the latest changes from main without creating a merge commit (in another words: re-apply commits from one branch to another) | `$ git rebase branch` // branch is the one that want to take the commits from|
|  Combines multiple commits into a single commit, often used to clean up commit history before merging or sharing. | `$ git rebase -i HEAD~n` // n is the number of commits and change "pick" to "squash" (see link)|
|  Shows the changes between your working directory and the staging area, or between two specific commits.(git pull=git fetch + git merge) | `$ git diff` |

## Link for details
+ git rebase VS merge: <https://youtu.be/zOnwgxiC0OA?si=7HLe_G4HW6AFkSfO>
+ git cherry-pick : <https://youtu.be/i657Bg_HAWI?si=rWeVRbFc4nKMJuJV>
+ git fetch VS git merge: <https://youtu.be/T13gDBXarj0?si=Pg8_UDI4Jx3GclTo>
+ git squash: <https://youtu.be/V5KrD7CmO4o?si=LwZIz7S6x7el_Qzw>
+ git reset VS git revert <https://youtu.be/LQ1NTTjI3ZU?si=gxWvlNtf90tElEk3>

**Comment by anomatomato** [comment]
## Git workflow (merge/rebase):

**Rebase**, when updating feature branch with latest changes from main branch 
```sh
git checkout feature-branch
git rebase main
```

**Merge**, when merging a feature branch into main branch
```sh
git checkout main
git merge feature-branch
```

:exclamation::exclamation::exclamation:Only rebase on local branches, use merge for public/shared ones
